iceoryx_posh/source/runtime/service_discovery.cpp

```diff
@@ -21,97 +21,47 @@ namespace iox
 {
 namespace runtime
 {
-cxx::expected<ServiceContainer, FindServiceError>
-ServiceDiscovery::findService(const cxx::optional<capro::IdString_t>& service,
-                              const cxx::optional<capro::IdString_t>& instance,
-                              const cxx::optional<capro::IdString_t>& event) noexcept
+ServiceContainer ServiceDiscovery::findService(const cxx::optional<capro::IdString_t>& service,
+                                               const cxx::optional<capro::IdString_t>& instance,
+                                               const cxx::optional<capro::IdString_t>& event) noexcept
 {
-    /// @todo #415 remove the string mapping, once the find call is done via shared memory
-    capro::IdString_t serviceString;
-    capro::IdString_t instanceString;
-    capro::IdString_t eventString;
-    bool isServiceWildcard = !service;
-    bool isInstanceWildcard = !instance;
-    bool isEventWildcard = !event;
+    ServiceContainer searchResult;
+    roudi::ServiceRegistry::ServiceDescriptionVector_t tempSearchResult;
 
-    if (!isServiceWildcard)
-    {
-        serviceString = service.value();
-    }
-    if (!isInstanceWildcard)
-    {
-        instanceString = instance.value();
-    }
-    if (!isEventWildcard)
-    {
-        eventString = event.value();
-    }
+    // Copy the new service registry if has changed
+    m_serviceRegistrySubscriber.take().and_then([&](popo::Sample<const roudi::ServiceRegistry>& serviceRegistrySample) {
+        m_serviceRegistry = *serviceRegistrySample;
+    });
 
-    IpcMessage sendBuffer;
-    sendBuffer << IpcMessageTypeToString(IpcMessageType::FIND_SERVICE) << PoshRuntime::getInstance().getInstanceName()
-               << cxx::convert::toString(isServiceWildcard) << serviceString
-               << cxx::convert::toString(isInstanceWildcard) << instanceString
-               << cxx::convert::toString(isEventWildcard) << eventString;
-
-    IpcMessage requestResponse;
-
-    if (!PoshRuntime::getInstance().sendRequestToRouDi(sendBuffer, requestResponse))
+    m_serviceRegistry.find(tempSearchResult, service, instance, event);
+    for (auto& service : tempSearchResult)
     {
-        LogError() << "Could not send FIND_SERVICE request to RouDi\n";
-        errorHandler(Error::kIPC_INTERFACE__REG_UNABLE_TO_WRITE_TO_ROUDI_CHANNEL, nullptr, ErrorLevel::MODERATE);
-        return cxx::error<FindServiceError>(FindServiceError::UNABLE_TO_WRITE_TO_ROUDI_CHANNEL);
+        searchResult.push_back(service.serviceDescription);
```

Comment:
We need to consider changing the registry o have the option of not getting the counter as well, then we do not have to transform the results. Can be combined with the lambda interface.

Reply:
I don't the user needs to be aware of how many `popo::Publisher`s are available.

Reply:
I just have an idea how to considerably optimize the registry and discovery with counters and more. But this is for the future.